Samples/BackgroundBlur/BackgroundBlur/Capture.h

```diff
@@ -0,0 +1,238 @@
+        pUnk->Release();
+    }
+    return hr;
+}
+
+
+struct ChooseDeviceParam
+{
+    ChooseDeviceParam() : ppDevices(NULL), count(0)
+    {
+    }
+    ~ChooseDeviceParam()
+    {
+        for (DWORD i = 0; i < count; i++)
+        {
+            SAFE_RELEASE(ppDevices[i]);
+        }
+        CoTaskMemFree(ppDevices);
+    }
+
+    IMFActivate** ppDevices;
+    UINT32      count;
+    UINT32      selection;
+};
+
+
+
+// CaptureManager class
+// Wraps the capture engine and implements the event callback.
+
+class CaptureManager
+{
+    // The event callback object.
+    class CaptureEngineCB : public IMFCaptureEngineOnEventCallback
+    {
+        long m_cRef;
+        HWND m_hwnd;
+
+    public:
+        CaptureEngineCB(HWND hwnd) : m_cRef(1), m_hwnd(hwnd), m_fSleeping(false), m_pManager(NULL) {}
+
+        // IUnknown
+        STDMETHODIMP QueryInterface(REFIID riid, void** ppv);
+        STDMETHODIMP_(ULONG) AddRef();
+        STDMETHODIMP_(ULONG) Release();
+
+        // IMFCaptureEngineOnEventCallback
+        STDMETHODIMP OnEvent(_In_ IMFMediaEvent* pEvent);
+
+        bool m_fSleeping;
+        CaptureManager* m_pManager;
+    };
+
+    HWND                    m_hwndEvent;
+    HWND                    m_hwndPreview;
+
+    IMFCaptureEngine* m_pEngine;
+    IMFCapturePreviewSink* m_pPreview;
+
+    CaptureEngineCB* m_pCallback;
+
+    bool                    m_bPreviewing;
+
+    UINT                    m_errorID;
+    HANDLE                  m_hEvent;
+    HANDLE                  m_hpwrRequest;
+    bool                    m_fPowerRequestSet;
+
+    CaptureManager(HWND hwnd) :
+        m_hwndEvent(hwnd), m_hwndPreview(NULL), m_pEngine(NULL), m_pPreview(NULL),
+        m_pCallback(NULL), m_bPreviewing(false), m_errorID(0), m_hEvent(NULL)
+        , m_hpwrRequest(INVALID_HANDLE_VALUE)
+        , m_fPowerRequestSet(false)
+    {
+        REASON_CONTEXT  pwrCtxt;
+
+        pwrCtxt.Version = POWER_REQUEST_CONTEXT_VERSION;
+        pwrCtxt.Flags = POWER_REQUEST_CONTEXT_SIMPLE_STRING;
+        pwrCtxt.Reason.SimpleReasonString = L"CaptureEngine is recording!";
+
+        m_hpwrRequest = PowerCreateRequest(&pwrCtxt);
+    }
+
+    void SetErrorID(HRESULT hr, UINT id)
+    {
+        m_errorID = SUCCEEDED(hr) ? 0 : id;
+    }
+
+    // Capture Engine Event Handlers
+    void OnCaptureEngineInitialized(HRESULT& hrStatus);
+    void OnPreviewStarted(HRESULT& hrStatus);
+    void OnPreviewStopped(HRESULT& hrStatus);
+    void WaitForResult()
+    {
+        WaitForSingleObject(m_hEvent, INFINITE);
+    }
+public:
+    ~CaptureManager()
+    {
+        DestroyCaptureEngine();
+    }
+
+    static HRESULT CreateInstance(HWND hwndEvent, CaptureManager** ppEngine)
+    {
+        HRESULT hr = S_OK;
+        *ppEngine = NULL;
+
+        CaptureManager* pEngine = new (std::nothrow) CaptureManager(hwndEvent);
+        if (pEngine == NULL)
+        {
+            hr = E_OUTOFMEMORY;
+            goto Exit;
```

Comment:
In general, we should avoid using goto as much as possible as a good coding practice, the logic here seems easy enough to replace it with some if else statements

Reply:
nvm, just noticed that this is open source code, might be a good idea to put open source code in a separate folder called 'external' or something

Reply:
Yeah I think do{} while(false) is actually a neat trick to avoid complicated nested if else statements (or duplicated return statements) and it can also be used as a way to encapsulate a few lines of code into a macro. goto on the other hand can be messy if not careful.